lib/BuildSystem/BuildSystem.cpp

```diff
@@ -3320,6 +3320,195 @@ class ArchiveTool : public Tool {
+      }
+    }
+    if (sharedLibInputs.empty()) {
+      ctx.error("SharedLibraryTool requires inputs to be specified");
+    }
+  }
+
+  virtual void configureOutputs(const ConfigureContext& ctx,
+                                const std::vector<Node*>& value) override {
+    ExternalCommand::configureOutputs(ctx, value);
+
+    for (const auto& output : getOutputs()) {
+      if (!output->isVirtual()) {
+        if (sharedLibName.empty()) {
+          sharedLibName = output->getName();
+        } else {
+          ctx.error("unexpected explicit output: " + output->getName());
+        }
+      }
+    }
+    if (sharedLibName.empty()) {
+      ctx.error("SharedLibraryTool requires the resulting shared library name "
+                "to be specified");
+    }
+  }
+
+  virtual bool configureAttribute(const ConfigureContext& ctx, StringRef name,
+                                  StringRef value) override {
+    if (name == "swiftExecutable") {
+      swiftExecutable = value;
+    } else if (name == "clangExecutable") {
+      clangExecutable = value;
+    } else if (name == "clang++Executable") {
+      clangPlusPlusExecutable = value;
+    } else if (name == "other-args") {
+      SmallVector<StringRef, 32> args;
+      StringRef(value).split(args, " ", /*MaxSplit=*/-1,
+                             /*KeepEmpty=*/false);
+      otherArgs = std::vector<std::string>(args.begin(), args.end());
+    } else if (name == "language") {
+      if (value != "C" && value != "C++" && value != "Swift") {
+        ctx.error("Unsupported language: '" + value +
+                  "' for shared libraries'.  Supported languages are C, C++, "
+                  "and Swift");
+      }
+      language = value;
+    }
+    return true;
+  }
+
+  virtual bool configureAttribute(const ConfigureContext& ctx, StringRef name,
+                                  ArrayRef<StringRef> values) override {
+    if (name == "other-args") {
+      otherArgs = std::vector<std::string>(values.begin(), values.end());
+    } else {
+      return ExternalCommand::configureAttribute(ctx, name, values);
+    }
+
+    return true;
+  }
+
+  std::vector<std::string> getArgs() const {
+    std::vector<std::string> args;
+
+    if (language == "Swift") {
+      args.push_back(swiftExecutable);
+      args.push_back("-emit-library");
+      args.insert(args.end(), sharedLibInputs.begin(), sharedLibInputs.end());
+      args.push_back("-o");
+      args.push_back(sharedLibName);
+      args.insert(args.end(), otherArgs.begin(), otherArgs.end());
+    } else {
+      if (language == "C") {
+        args.push_back(clangExecutable);
+      } else if (language == "C++") {
+        args.push_back(clangPlusPlusExecutable);
+      }
+      args.insert(args.end(), sharedLibInputs.begin(), sharedLibInputs.end());
+      args.push_back("-o");
+      args.push_back(sharedLibName);
+      args.insert(args.end(), otherArgs.begin(), otherArgs.end());
+
+#if defined(_WIN32)
```

Comment:
Do we care about cross compilation? If so, I'll add a target parameter and base the options added on that.

Reply:
Yes, please do so. LLVM philosophy :)